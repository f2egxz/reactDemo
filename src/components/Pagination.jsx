//-*- mode: react -*-
//-*- coding: utf-8 -*-

/**
 * Pagination.jsx
 *
 * 分页组件
 *
 * -  TODO 显示所有页数的下拉菜单
 * -  TODO 隐藏部分页码的下拉菜单
 *
 * @author fe.gxz
 */

import React,{Component}  from 'react'
import { classnames }     from 'util/view-extra'
import Styles             from 'style/Pagination.css'

/**
 * Pagination
 *
 * 分页组件
 *
 * @param {number} total
 * @param {number} pageSize
 * @param {number} current
 * @param {number} maxShowPage
 * @param {func} onChange
 *
 *
 */
class Pagination extends Component {
  constructor(){
    super()
  }
  render(){
    const {total,pageSize,current,maxShowPage} = this.props;
    const pageNum = Math.ceil(total/pageSize);
    if(pageNum>maxShowPage){
      return (
        <ul className={classnames(Styles.defaultUl)}>
          <li className={classnames(Styles.defaultLi,current===1?Styles.lose:'')}
              onClick={this.handleBack.bind(this)}
          >	&lt;</li>
          <li className={this.MapCurrentToStyle (current,1)}
              onClick={this.handlePage.bind(this)}
          >1</li>
          {current > 3 ?
            <li className={classnames(Styles.defaultLi,Styles.more,Styles.moreLt)}
                onClick={this.handleBack.bind(this)}
            ></li>
            :''
          }
          <li className={this.MapCurrentToStyle (current,2,pageNum)}
              onClick={this.handlePage.bind(this)}
          > {current > 3&&current < pageNum-2 ?current-1:current < 4 ? 2 : pageNum-3}</li>
          <li className={this.MapCurrentToStyle (current,3,pageNum)}
              onClick={this.handlePage.bind(this)}
          > {current > 3&&current < pageNum-2 ?current  :current < 4 ? 3 : pageNum-2}</li>
          <li className={this.MapCurrentToStyle (current,4,pageNum)}
              onClick={this.handlePage.bind(this)}
          > {current > 3&&current < pageNum-2 ?current+1:current < 4 ? 4 : pageNum-1}</li>
          {current < pageNum-2 ?
            <li  className={classnames(Styles.defaultLi,Styles.more,Styles.moreRt)}
                                      onClick={this.handleGo.bind(this)}
            ></li>
            :''
          }
          <li className={this.MapCurrentToStyle (current,pageNum,pageNum)}
              onClick={this.handlePage.bind(this)}
          >{pageNum}</li>
          <li className={classnames(Styles.defaultLi,current===pageNum?Styles.lose:'')}
              onClick={this.handleGo.bind(this)}
          >	&gt;</li>
        </ul>
      )
    }else{
      let page = [];
      for(let i=1;i<=pageNum;i++){
        page.push(<li key={i} className={this.MapCurrentToStyle (current,i)} onClick={this.handlePage.bind(this)}>{i}</li>)
      }
      return (
        <ul className={classnames(Styles.defaultUl)}>
          <li className={classnames(Styles.defaultLi,current===1?Styles.lose:'')}
              onClick={this.handleBack.bind(this)}
          >	&lt;</li>
          {page}
          <li className={classnames(Styles.defaultLi,current===pageNum?Styles.lose:'')}
              onClick={this.handleGo.bind(this)}
          >	&gt;</li>
        </ul>
      )
    }
  }

  /**
   * 根据当前选中页确定样式
   * @param current
   * @param num
   * @param pageNum
   * @return {*}
   * @constructor
   */
  MapCurrentToStyle(current,num,pageNum=0){
    if(pageNum===0){
      //不省略页码时
      if(current===num){
        return classnames(Styles.defaultLi,Styles.current);
      }else {
        return classnames(Styles.defaultLi)
      }
    }else {
      //有页码隐藏时
      if(current===num&&(num<4||num===pageNum)){
        return classnames(Styles.defaultLi,Styles.current);
      }else if(current > 3 && current < pageNum-2 && num===3){
        return classnames(Styles.defaultLi,Styles.current)
      }else if(current === pageNum-2 && current>3 && num===3){
        return classnames(Styles.defaultLi,Styles.current)
      }else if(current === pageNum-1 && current>3 && num===4){
        return classnames(Styles.defaultLi,Styles.current)
      }else {
        return classnames(Styles.defaultLi)
      }
    }
  }
  handleBack(){
    const {current,onChange} = this.props
    current>1? onChange(current-1):''
  }
  handleGo(){
    const {total,pageSize,current,onChange} = this.props
    const pageNum = total/pageSize;
    current<pageNum? onChange(current+1):''
  }
  handlePage(e){
    const page =parseInt( e.target.innerText);
    const {current,onChange} = this.props;
    current!==page? onChange(page):''
  }
}

Pagination.defaultProps = {
  pageSize: 10,
  maxShowPage: 8
}

Pagination.PropTypes = {
  onChange:React.PropTypes.func.isRequired,
  total:React.PropTypes.number.isRequired,
  pageSize:React.PropTypes.number,
  current:React.PropTypes.number.isRequired,
  maxShowPage:React.PropTypes.number
}

export default Pagination
