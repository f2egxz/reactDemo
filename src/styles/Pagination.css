/*-*- mode: css -*-*/
/*-*- coding: utf-8 -*-*/

/**
 * Pagination.css
 *
 * 分页组件
 *
 * @author fe.gxz
 */

@value themeColor:          #108ee9;

.defaultUl:before,  .defaultUl:after {
  content: "";
  display: table;
}
.defaultUl:after {
  clear: both;
}
.defaultUl {
  list-style: none;
  margin:0;
  padding: 0;
  zoom: 1;/*IE678*/
}

.defaultLi {
  cursor: pointer;
  color: rgba(0, 0, 0, 0.65);
  border-radius: 4px;
  list-style: none;
  min-width: 28px;
  height: 28px;
  line-height: 28px;
  float: left;
  text-align: center;
  transition: all 0.3s ease;
  display: inline-block;
  border: 1px solid #d9d9d9;
  background-color: #fff;
  margin-right: 5px;
  -moz-user-select:none;
  user-select:none;
  &:hover {
     border: 1px solid themeColor;
     color: themeColor;
   }
}

.more {
  border: none;
  font-size: 12px;
  &:hover {
     border: none;
   }
}

.more::after {
  content: '▪▪▪';
}

.moreLt:hover::after {
  content: '<<';
  font-size: 16px;
}

.moreRt:hover::after {
  content: '>>';
  font-size: 16px;
}

.current {
  background-color: themeColor;
  border: 1px solid themeColor;
  color: #fff;
  &:hover {
    color:#fff
   }
}

.lose {
  border: 1px solid #d9d9d9;
  color: #d9d9d9;
  background-color: #fff;
  cursor: not-allowed;
  &:hover {
     cursor: not-allowed;
     border: 1px solid #d9d9d9;
     color: #d9d9d9;
   }
}


