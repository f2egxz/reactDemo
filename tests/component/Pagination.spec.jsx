import React from 'react'
import Pagination from 'component/Pagination.jsx'
import render from 'react-test-renderer'

test('Pagination',()=>{
  const PagChange= jest.fn()
  const component = render.create(
    <Pagination total={322} current={5} onChange={PagChange}/>
  )
  let PagString = component.toJSON();
  expect(PagString).toMatchSnapshot();
})
