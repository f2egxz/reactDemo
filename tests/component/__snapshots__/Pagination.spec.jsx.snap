exports[`test Pagination 1`] = `
<ul
  className="defaultUl">
  <li
    className="defaultLi "
    onClick={[Function]}>
     <
  </li>
  <li
    className="defaultLi"
    onClick={[Function]}>
    1
  </li>
  <li
    className="defaultLi more moreLt"
    onClick={[Function]} />
  <li
    className="defaultLi"
    onClick={[Function]}>
     
    4
  </li>
  <li
    className="defaultLi current"
    onClick={[Function]}>
     
    5
  </li>
  <li
    className="defaultLi"
    onClick={[Function]}>
     
    6
  </li>
  <li
    className="defaultLi more moreRt"
    onClick={[Function]} />
  <li
    className="defaultLi"
    onClick={[Function]}>
    33
  </li>
  <li
    className="defaultLi "
    onClick={[Function]}>
     >
  </li>
</ul>
`;
